### ISSUE TITLE
Page with same name as site path prefix yields 404

### ISSUE THREAD

**Issue by mxxk** [gatsbyjs/gatsby#15334]
## Description

Gatsby is not able to show a page whose path matches the site path prefix. If gatsby-config.js has

```js
module.exports = {
  pathPrefix: `/prefix`,
}
```

and there is a page `src/pages/prefix.js`, then navigating to this page (`/prefix/prefix/`) yields a 404 Not Found. The Developer Tools console shows a request to the `/prefix/page-data//page-data.json`, whereas it should be requesting `/prefix/page-data/prefix/page-data.json`.

This issue occurs only when:

1. The page path matches the site prefix exactly.
2. The site is built with path prefixes.

### Steps to reproduce

1. Set a path prefix on a Gatsby site.

2. Create a page whose path matches the site prefix.

3. Build site with path prefixes:

    ```
    gatsby build --prefix-paths
    ```
    (When the site it built with regular `gatsby build` or `gatsby develop`, this issue does not come up.)

See [example site](https://mxxk.github.io/test-gatsby-path-prefix-navigation/) and [GitHub project](https://github.com/mxxk/test-gatsby-path-prefix-navigation).

### Expected result

Navigating to [Page 2](https://mxxk.github.io/test-gatsby-path-prefix-navigation/test-gatsby-path-prefix-navigation/) would show the sweet, sweet, greeting we all know and love:

![image](https://user-images.githubusercontent.com/2456381/60550939-76b8b500-9cde-11e9-9f83-2b401965fec2.png)

### Actual result

No greeting! 😢 Navigating to [Page 2](https://mxxk.github.io/test-gatsby-path-prefix-navigation/test-gatsby-path-prefix-navigation/) yields a 404:

![image](https://user-images.githubusercontent.com/2456381/60550662-8daad780-9cdd-11e9-9c7c-8a410db16849.png)

### Environment

```
$ gatsby info --clipboard

  System:
    OS: macOS High Sierra 10.13.6
    CPU: (8) x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
    Yarn: 1.16.0 - ~/.yarn/bin/yarn
    npm: 6.4.1 - ~/.nvm/versions/node/v10.15.3/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 75.0.3770.100
    Firefox: 67.0.4
    Safari: 12.1.1
```

**Comment by m-allanson** [comment]
Thanks for the repro! I can confirm I'm seeing the same issue. One thing to note is that a pathPrefix should always start with a leading slash: `pathPrefix: "/prefix"`. 

However that doesn't fix the issue. If you or anyone else is up to investigating this, I'm taking a guess that the issue might be around here: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/loader.js

One way to move this issue along could be to add a failing e2e test to the path-prefix tests: https://github.com/gatsbyjs/gatsby/blob/master/e2e-tests/README.md#L21

**Comment by m-allanson** [comment]
cc @wardpeet if you have any thoughts on this?

**Comment by wardpeet** [comment]
no, but I'll investigate 🕵

**Comment by henrikwirth** [comment]
I have the same problem. In my case the pathPrefix is `site` and the page is `site18` (example).
Then if I navigate to `example.com/site/site18` it is trying to fetch the data from `example.com/site/18` instead. So it cuts away the prefix from the actual page url.

Maybe its stripped in a wrong way here: https://github.com/gatsbyjs/gatsby/blob/63f8a6c1de2ae2af512948bcf265a058905bb693/packages/gatsby/cache-dir/loader.js#L8-L12

**Comment by nkuehn** [comment]
I can confirm the issue for any pages or paths containing the prefix,  e.g.  `/site/site-variant/` when the path prefix is `/site`.   Can also confirm that it's an actual hindrance.

**Comment by bchehraz** [comment]
So I'm not sure if I'm having the same exact issue here, but here's what I'm currently working on. I got client side routing with "/" and "/someOtherPage" being accessible by the client. And if they try to access a page they don't yet have access to, it will redirect them to "/app/login" where they have to authenticate with the server. And I have a path prefix because I'd like to put this on GitHub Pages, but since putting it on there, I've been getting 404 errors with loading in the data. 

I am using a Router & Redirect from @reach/router in the following context:

```
<Router>
  {this.state.token && <Redirect from="/app/login" to="/app/events" exact noThrow />}
  {this.state.token && <Redirect from="/" to="/app/events" exact noThrow />}

  <Events path="/app/events" />
  <PrivateRoute path="/app/bookings" component={Bookings} />
  <Login signUp={false} path="/app/login" />
  <Login signUp path="/app/sign-up" />
</Router>
```

PrivateRoute just takes the path and passes the specific component for that page if they are authenticated.

And in my gatsby-node.js...

```
exports.onCreatePage = async ({ page, actions }) => {
  const { createPage } = actions;

  // page.matchPath is a special key that's used for matching pages
  // only on the client
  if (page.path.match(/^\/app/)) {
    page.matchPath = `/app/*`;

    createPage(page);
  }
}
```

It's pretty much based on the logic in one of the Gatsby starter examples called Simple Auth (https://github.com/gatsbyjs/gatsby/tree/master/examples/simple-auth)

**Comment by bchehraz** [comment]
So, I'm not exactly sure if this is related or not, but I'm currently working on a project where I'm attempting to use client-only routes with authentication to access the hidden pages. After adding in pathPrefix to use GitHub Pages with the project, I started to get that 404 response discussed above.

I've created an example project which is based on the Simple Auth gatsby project (https://github.com/gatsbyjs/gatsby/tree/master/examples/simple-auth) which is demonstrating client routing. Upon adding the pathPrefix "/simple-auth/" and uploading the same project to GitHub, it's now getting the same error.

[Demo project here](https://bchehraz.github.io/simple-auth/)

What's supposed to happen is whenever you go to Profile or Details page, it should redirect to the login page, but if you look at the console logs, there's that 404 error.
